lib/tree.rb

```diff
@@ -16,40 +16,129 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # def add(key, value)
+  #   if @root.nil?
+  #     @root = TreeNode.new(key, value)
+  #   else 
+  #     current = @root
+
+  #     while true
+  #       if key < current.key
+  #         if !current.left.nil?
+  #           current = current.left
+  #         else
+  #           current.left = TreeNode.new(key, value)
+  #           return
+  #         end
+  #       else
+  #         if !current.right.nil?
+  #           current = current.right
+  #         else
+  #           current.right = TreeNode.new(key, value)
+  #           return
+  #         end
+  #       end
+  #     end
+  #   end
+  # end
+
+  def add_helper(current_node, key, value)
+    return TreeNode.new(key, value) if current_node.nil?
+
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value)
+    end
+
+    return current_node
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
   def add(key, value)
```

Comment:
This method is **recursive** so it's going to have some space complexity.  What do you think it is?